### ISSUE TITLE
Incorrect hint for code involving ``iso`` function from Control.Lens.Iso

### ISSUE THREAD

**Issue by kozross** [ndmitchell/hlint#240]
In the following code:

```
{-# Language RankNTypes #-}
{-# Language TemplateHaskell #-}
module Demo where

import qualified Data.Set as Set
import Data.Set (Set)
import Control.Lens

data Atom = Pos { i :: Integer}
          | Neg { i :: Integer}
          deriving (Eq, Ord, Show, Read)

negatingAtom :: Iso' Atom Atom
negatingAtom = iso go go
  where go (Pos x) = Neg x
        go (Neg x) = Pos x

newtype Conjunction a = Conjunction (Set a)
  deriving (Eq, Ord, Read, Show)

newtype Disjunction a = Disjunction (Set a)
  deriving (Eq, Ord, Read, Show)

makePrisms ''Disjunction
makePrisms ''Conjunction


negatingClause :: Iso' (Conjunction Atom) (Disjunction Atom)
negatingClause = liftClause negatingAtom

type CNF = Conjunction (Disjunction Atom)
type DNF = Disjunction (Conjunction Atom)


-- Helper stuff

liftSetIso :: (Ord a, Ord b) => AnIso' a b -> Iso' (Set a) (Set b)
liftSetIso x = withIso x $ \sa bt -> iso (Set.map sa) (Set.map bt)

liftClause :: (Ord a, Ord b) => AnIso' a b -> Iso' (Conjunction a) (Disjunction b)
liftClause x = _Conjunction . liftSetIso x . from _Disjunction

negatingForm :: Iso' CNF DNF
negatingForm = liftClause (from negatingClause)

```

HLint gives a suggestion to replace ``\sa bt -> iso (Set.map sa) (Set.map bt)`` with ``iso `on` Set.map``. However, this won't typecheck in this case (and indeed, in general), as ``sa`` and ``bt`` do not agree on their types, which ``on`` expects.

**Comment by ndmitchell** [comment]
Good point! Looks like this hint needs to be removed.

**Comment by ndmitchell** [comment]
Removed and released as hlint-1.9.33.
